### PR TITLE
Line chart value replacement for Mongo DB

### LINKED ISSUE #16170: Line Chart not working for mongodb when using Replace Missing Values with nothing or zero
**Describe the bug**
After updating metabase to v0.39 my Line charts that has X-axis scale set to Timeseries started to present all zero values even if in the data visualization the data is correct. I noticed that it only occurs on data stored in mongo db, I have others charts that gets data from other sources and it was not affected by the problem.

**Logs**
In the server side there is any errors
I noticed the following warnings in the browser console:
image

**To Reproduce**
1. Start a new question with data stored in mongo db
2. Sumarize the data from a date column
3. Change the chart type to Line
4. Select "Replace Missing Values" to "Zero" or "Nothing"
5. Every point in the chart goes to zero

**Expected behavior**
The points in the chart that is not zero have to mantain their positions

**Screenshots**
Replace Missing Values = Line Interpolated:
image

Replace Missing Values = Zero:
image

**Information about your Metabase Installation:**
```
{
  "browser-info": {
    "language": "en-US",
    "platform": "Win32",
    "userAgent": "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36",
    "vendor": "Google Inc."
  },
  "system-info": {
    "file.encoding": "UTF-8",
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.11+9",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.11",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.11+9",
    "os.name": "Linux",
    "os.version": "4.15.0-143-generic",
    "user.language": "en",
    "user.timezone": "America/Sao_Paulo"
  },
  "metabase-info": {
    "databases": [
      "mongo",
      "sqlserver",
      "postgres"
    ],
    "hosting-env": "unknown",
    "application-database": "postgres",
    "application-database-details": {
      "database": {
        "name": "PostgreSQL",
        "version": "9.6.20"
      },
      "jdbc-driver": {
        "name": "PostgreSQL JDBC Driver",
        "version": "42.2.18"
      }
    },
    "run-mode": "prod",
    "version": {
      "date": "2021-05-17",
      "tag": "v0.39.2",
      "branch": "release-x.39.x",
      "hash": "b0d9436"
    },
    "settings": {
      "report-timezone": "America/Sao_Paulo"
    }
  }
}
```

- Your browser and the version: Chrome 90.0
- Your operating system: Windows 7 x64
- Your databases: MongoDB 3.6
- Metabase version: v0.39.2
- Metabase hosting environment: Docker
- Metabase internal database: Postgres

**Severity**
It is a big problem because I lost some important charts for the company that started to present all zero values.

### PR DESCRIPTION
**Pull request by losrebellos**
#### Steps to replicate for MongoDB

1. create a question on MongoDB [repro](https://github.com/metabase/success-engineering/tree/master/repro/mongo) data
2. summarize by `count` and group by `created at`
3. filter `created at` to be before `31/01/2017`

<details>
  <summary>The received data will contains `type/Instant` with date of the form `2016-01-01T01:00:00+01:00`</summary>

```json
{
  "data":{
    "rows":[
      [
        "2016-01-01T01:00:00+01:00",
        744
      ],
      [
        "2017-01-01T01:00:00+01:00",
        3610
      ],
      [
        "2018-01-01T01:00:00+01:00",
        5834
      ],
      [
        "2019-01-01T01:00:00+01:00",
        6578
      ],
      [
        "2020-01-01T01:00:00+01:00",
        1994
      ]
    ],
    "cols":[
      {
        "description":null,
        "semantic_type":"type/CreationTimestamp",
        "table_id":5,
        "coercion_strategy":null,
        "unit":"month",
        "name":"created_at",
        "settings":null,
        "source":"breakout",
        "field_ref":[
          "field",
          39,
          {
            "temporal-unit":"month"
          }
        ],
        "effective_type":"type/Instant",
        "parent_id":null,
        "id":39,
        "visibility_type":"normal",
        "display_name":"Created At",
        "fingerprint":{
          "global":{
            "distinct-count":10000,
            "nil%":0.0
          },
          "type":{
            "type/DateTime":{
              "earliest":"2016-01-01T02:36:07.53Z",
              "latest":"2020-01-31T23:28:11.385Z"
            }
          }
        },
        "base_type":"type/Instant"
      },
      {
        "base_type":"type/BigInteger",
        "semantic_type":"type/Quantity",
        "name":"count",
        "display_name":"Count",
        "source":"aggregation",
        "field_ref":[
          "aggregation",
          0
        ],
        "effective_type":"type/BigInteger"
      }
    ],
    "native_form":{
      "projections":[
        "created_at~~~month",
        "count"
      ],
      "query":[
        {
          "$group":{
            "_id":{
              "created_at~~~month":{
                "$let":{
                  "vars":{
                    "parts":{
                      "$dateToParts":{
                        "date":"$created_at"
                      }
                    }
                  },
                  "in":{
                    "$dateFromParts":{
                      "year":"$$parts.year",
                      "month":"$$parts.month"
                    }
                  }
                }
              }
            },
            "count":{
              "$sum":1
            }
          }
        },
        {
          "$sort":{
            "_id":1
          }
        },
        {
          "$project":{
            "_id":false,
            "created_at~~~month":"$_id.created_at~~~month",
            "count":true
          }
        },
        {
          "$sort":{
            "created_at~~~month":1
          }
        }
      ],
      "collection":"orders",
      "mbql?":true
    },
    "results_timezone":"Europe/Paris",
    "results_metadata":{
      "columns":[
        {
          "description":null,
          "semantic_type":"type/CreationTimestamp",
          "coercion_strategy":null,
          "unit":"month",
          "name":"created_at",
          "settings":null,
          "field_ref":[
            "field",
            39,
            {
              "temporal-unit":"month"
            }
          ],
          "effective_type":"type/Instant",
          "id":39,
          "display_name":"Created At",
          "fingerprint":{
            "global":{
              "distinct-count":10000,
              "nil%":0.0
            },
            "type":{
              "type/DateTime":{
                "earliest":"2016-01-01T02:36:07.53Z",
                "latest":"2020-01-31T23:28:11.385Z"
              }
            }
          },
          "base_type":"type/Instant"
        },
        {
          "display_name":"Count",
          "semantic_type":"type/Quantity",
          "field_ref":[
            "aggregation",
            0
          ],
          "name":"count",
          "base_type":"type/BigInteger",
          "effective_type":"type/BigInteger",
          "fingerprint":{
            "global":{
              "distinct-count":5,
              "nil%":0.0
            },
            "type":{
              "type/Number":{
                "min":744.0,
                "q1":1681.5,
                "q3":6020.0,
                "max":6578.0,
                "sd":2473.8528654711863,
                "avg":3752.0
              }
            }
          }
        }
      ]
    },
    "insights":[
      {
        "previous-value":null,
        "unit":"month",
        "offset":-22522.851137442918,
        "last-change":null,
        "col":"count",
        "slope":1.4986967189588587,
        "last-value":1994,
        "best-fit":[
          "+",
          -260781.80100388598,
          [
            "*",
            27072.42520229,
            [
              "log",
              "x"
            ]
          ]
        ]
      }
    ]
  },
  "database_id":2,
  "started_at":"2022-03-15T16:53:20.757287+01:00",
  "json_query":{
    "constraints":{
      "max-results":10000,
      "max-results-bare-rows":2000
    },
    "type":"query",
    "middleware":{
      "js-int-to-string?":true,
      "ignore-cached-results?":false,
      "process-viz-settings?":false
    },
    "database":2,
    "query":{
      "source-table":5,
      "aggregation":[
        [
          "count"
        ]
      ],
      "breakout":[
        [
          "field",
          39,
          {
            "temporal-unit":"month"
          }
        ]
      ]
    },
    "async?":true,
    "cache-ttl":null
  },
  "average_execution_time":null,
  "status":"completed",
  "context":"question",
  "row_count":5,
  "running_time":651
}
```
</details>

#### Steps with default data sample

1. create a question on MongoDB `repro` data
2. summarize by `count` and group by `created at`
3. filter `count` to be greater than `550`

<details>
  <summary>The received data will contains `type/DateTime` with date of the form `2019-01-01T00:00:00+01:00`
  </summary>

```json
{
  "data": {
    "rows": [
      ["2019-01-01T00:00:00+01:00", 584],
      ["2019-03-01T00:00:00+01:00", 573],
      ["2019-05-01T00:00:00+02:00", 561],
      ["2019-07-01T00:00:00+02:00", 556],
      ["2019-08-01T00:00:00+02:00", 566],
      ["2020-01-01T00:00:00+01:00", 580]
    ],
    "cols": [
      {
        "description": "The date and time an order was submitted.",
        "semantic_type": "type/CreationTimestamp",
        "table_id": 2,
        "coercion_strategy": null,
        "unit": "month",
        "name": "CREATED_AT",
        "settings": null,
        "source": "fields",
        "field_ref": [
          "field",
          12,
          {
            "temporal-unit": "default"
          }
        ],
        "effective_type": "type/DateTime",
        "parent_id": null,
        "id": 12,
        "visibility_type": "normal",
        "display_name": "Created At",
        "fingerprint": {
          "global": {
            "distinct-count": 9998,
            "nil%": 0.0
          },
          "type": {
            "type/DateTime": {
              "earliest": "2016-04-30T18:56:13.352Z",
              "latest": "2020-04-19T14:07:15.657Z"
            }
          }
        },
        "base_type": "type/DateTime"
      },
      {
        "base_type": "type/BigInteger",
        "semantic_type": "type/Quantity",
        "name": "count",
        "display_name": "Count",
        "source": "fields",
        "field_ref": [
          "field",
          "count",
          {
            "base-type": "type/BigInteger"
          }
        ],
        "effective_type": "type/BigInteger"
      }
    ],
    "native_form": {
      "query": "SELECT \"source\".\"CREATED_AT\" AS \"CREATED_AT\", \"source\".\"count\" AS \"count\" FROM (SELECT parsedatetime(formatdatetime(\"PUBLIC\".\"ORDERS\".\"CREATED_AT\", 'yyyyMM'), 'yyyyMM') AS \"CREATED_AT\", count(*) AS \"count\" FROM \"PUBLIC\".\"ORDERS\" GROUP BY parsedatetime(formatdatetime(\"PUBLIC\".\"ORDERS\".\"CREATED_AT\", 'yyyyMM'), 'yyyyMM') ORDER BY parsedatetime(formatdatetime(\"PUBLIC\".\"ORDERS\".\"CREATED_AT\", 'yyyyMM'), 'yyyyMM') ASC) \"source\" WHERE \"source\".\"count\" > 550 LIMIT 2000",
      "params": null
    },
    "results_timezone": "Europe/Paris",
    "results_metadata": {
      "columns": [
        {
          "description": "The date and time an order was submitted.",
          "semantic_type": "type/CreationTimestamp",
          "coercion_strategy": null,
          "unit": "month",
          "name": "CREATED_AT",
          "settings": null,
          "field_ref": [
            "field",
            12,
            {
              "temporal-unit": "default"
            }
          ],
          "effective_type": "type/DateTime",
          "id": 12,
          "display_name": "Created At",
          "fingerprint": {
            "global": {
              "distinct-count": 9998,
              "nil%": 0.0
            },
            "type": {
              "type/DateTime": {
                "earliest": "2016-04-30T18:56:13.352Z",
                "latest": "2020-04-19T14:07:15.657Z"
              }
            }
          },
          "base_type": "type/DateTime"
        },
        {
          "display_name": "Count",
          "semantic_type": "type/Quantity",
          "field_ref": [
            "field",
            "count",
            {
              "base-type": "type/BigInteger"
            }
          ],
          "name": "count",
          "base_type": "type/BigInteger",
          "effective_type": "type/BigInteger",
          "fingerprint": {
            "global": {
              "distinct-count": 6,
              "nil%": 0.0
            },
            "type": {
              "type/Number": {
                "min": 556.0,
                "q1": 561.0,
                "q3": 580.0,
                "max": 584.0,
                "sd": 10.936178491593852,
                "avg": 570.0
              }
            }
          }
        }
      ]
    },
    "insights": [
      {
        "previous-value": null,
        "unit": "month",
        "offset": 708.7431188478355,
        "last-change": null,
        "col": "count",
        "slope": -0.007685277698890775,
        "last-value": 580,
        "best-fit": [
          "+",
          1979.860583268822,
          ["*", -143.84788499045666, ["log", "x"]]
        ]
      }
    ]
  },
  "database_id": 1,
  "started_at": "2022-03-15T16:49:19.670638+01:00",
  "json_query": {
    "constraints": {
      "max-results": 10000,
      "max-results-bare-rows": 2000
    },
    "type": "query",
    "middleware": {
      "js-int-to-string?": true,
      "ignore-cached-results?": false,
      "process-viz-settings?": false
    },
    "database": 1,
    "query": {
      "source-query": {
        "source-table": 2,
        "aggregation": [["count"]],
        "breakout": [
          [
            "field",
            12,
            {
              "temporal-unit": "month"
            }
          ]
        ]
      },
      "filter": [
        ">",
        [
          "field",
          "count",
          {
            "base-type": "type/Integer"
          }
        ],
        550
      ]
    },
    "async?": true,
    "cache-ttl": null
  },
  "average_execution_time": null,
  "status": "completed",
  "context": "question",
  "row_count": 6,
  "running_time": 93
}
```
</details>

#### Differences

The provided dates format have a time difference: `...T01...` for MongoDB with `type/Instant` versus `...T00` for `type/DateTime`.

### Results

1. When updating the data with replacement values, [moment.toISOString](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/fill_data.js#L92) is used as a key string to create a new [map](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/fill_data.js#L20-L59)
2. When creating the `ticks` using the [ticksForRange](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/timeseriesScale.js#L86-L107) method, the [ticks are reseted](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/timeseriesScale.js#L91) to the beginning of their `unit` (month in our examples)
3. `moment.startOf` will use the __timezone__, generated keys will be:
    -  for `2016-01-01T01:00:00+01:00` => `2016-01-01T00:00:00.000Z`
    -  for `2019-01-01T00:00:00+01:00` => `2018-12-31T23:00:00.000Z`
4. Because of that time difference, when the [fill data](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/fill_data.js#L61-L111) are created to replace the missing values, the key string values used in the map and generated by `moment.toISOString` is not matching the key string values generated by the [ticksForRange](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/visualizations/lib/timeseriesScale.js#L86-L107) method

So in effect, it is skipping all original values (`date` + `value`) because of the `type/Instant` difference.

### Notes

It has been pointed out that `type/Instant` might not be mapped or used properly in connection with [isa](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/lib/types.js#L15).
After some investigations in [schema_metadata.js](https://github.com/metabase/metabase/blob/master/frontend/src/metabase/lib/schema_metadata.js#L108-L114), it seems that `type/Instant` is mapped properly to `type/Temporal` and the logic applied seems to be the same.

#### MongoDB

image

#### Default sample

image

#### MongoDB

image

#### Default sample

image

### CODE REVIEW

File: frontend/src/metabase/visualizations/lib/renderer_utils.js [hunk 1]
```diff
@@ -161,7 +161,9 @@ export function getDatas({ settings, series }, warn) {
 
     return rows.map(row => {
       const [x, ...rest] = row;
-      const newRow = [parseXValue(x, parseOptions, warn), ...rest];
+      const { unit } = parseOptions;
+      const xValue = parseXValue(x, parseOptions, warn);
+      const newRow = [xValue.startOf(unit), ...rest];
```
Comment (losrebellos) [inline frontend/src/metabase/visualizations/lib/renderer_utils.js:166]: By using `xValue.startOf(unit)`, the date will be reseted to the beginning of the `unit` (month in the examples) and key strings will be retrieved correctly in the data map with filled missing values:
-  with `type/Instant` for `2016-01-01T01:00:00+01:00` => `2015-12-31T23:00:00.000Z`
-  with `type/DateTime` for `2019-01-01T00:00:00+01:00` => `2018-12-31T23:00:00.000Z`

File: frontend/src/metabase/visualizations/lib/renderer_utils.js [hunk 2]
```diff
@@ -161,7 +161,13 @@ export function getDatas({ settings, series }, warn) {
 
     return rows.map(row => {
       const [x, ...rest] = row;
-      const newRow = [parseXValue(x, parseOptions, warn), ...rest];
+      const { unit } = parseOptions;
+      const xValue = parseXValue(x, parseOptions, warn);
+      const formattedXValue =
+        xValue && unit && typeof xValue.startOf === "function"
+          ? xValue.startOf(unit)
```
Comment (losrebellos) [inline frontend/src/metabase/visualizations/lib/renderer_utils.js:168]: By using `xValue.startOf(unit)`, the date will be reseted to the beginning of the `unit` (month in the examples) and key strings will be retrieved correctly in the data map with filled missing values:
-  with `type/Instant` for `2016-01-01T01:00:00+01:00` => `2015-12-31T23:00:00.000Z`
-  with `type/DateTime` for `2019-01-01T00:00:00+01:00` => `2018-12-31T23:00:00.000Z`

**Review by flamber — COMMENTED** [review]
This works for Mongo. The code makes me a little nervous, since it's a core change and seems to mess with timezones. Would love to get some timezone-frontend eyes on it.

But still seeing strange behavior on Druid, which is likely because it changes it to type/Text.
image
image
image

**Comment by losrebellos** [comment]
@flamber

Completely agree, hence the full description above:
- It is a core change so __we need__ some more reviews
- @nemanjaglumac will write some tests regarding that for MongoDB
- `Druid` and `type/Text` seems to be another completely separated issue

**Review by kulyk — APPROVED** [review]
LGTM

**Review by daltojohnso — APPROVED** [review]

**Comment by nemanjaglumac** [comment]
There was actually a repro written for this issue long time ago. You can see the `.Reproduced` label when you navigate to #16170 . Anyway, there is a faster and more reliable way to write that repro and I'm about to publish it.

Can we merge this fix now?
I tested it locally and it seems to be working + two FE devs approved.

**Comment by nemanjaglumac** [comment]
@losrebellos since this PR was still open, I pushed the repro directly to it, rather than you having to rebase your changes after I merge it in `master` or me having to be blocked by this PR.

Feel free to merge them together.

**Pull request merged**